### PR TITLE
Dedx

### PR DESCRIPTION
**Pull request by kjc5697**

### CODE REVIEW

File: retro/discrete_muon_kernels.py [hunk 1]
```diff
@@ -118,3 +120,109 @@ def const_energy_loss_muon(hypo_params, limits=None, dt=1.0):
...
+        pinfo_gen = np.array(
+            [[hypo_params.t,
+              hypo_params.x,
+              hypo_params.y,
+              hypo_params.z,
+              0,
+              0,
+              0,
+              0]],
+            dtype=np.float64
```
Comment (jllanfranchi) [inline retro/discrete_muon_kernels.py]: Docstring says float32 ... and I use float64 elsewhere but I don't see why we need double precision here, so convert too float32 (unless you disagree that's sufficient)

File: retro/discrete_muon_kernels.py [hunk 2]
```diff
@@ -118,3 +120,109 @@ def const_energy_loss_muon(hypo_params, limits=None, dt=1.0):
...
+        )
+        return pinfo_gen        
+
+    # create spline
+    with open(join(dirname(abspath(__file__)), 'dedx_total_e.csv'), 'rb') as csvfile:
```
Comment (jllanfranchi) [inline retro/discrete_muon_kernels.py]: Every time we call your function it opens and reads the same info from the same file. The slowest thing you can do on a computer besides querying the internet is reading from disk. Move this outside the function so it only happens once when the function is imported the first time. (Include the creation of the spline object)

File: retro/discrete_muon_kernels.py [hunk 3]
```diff
@@ -118,3 +120,109 @@ def const_energy_loss_muon(hypo_params, limits=None, dt=1.0):
...
+    dir_z = -math.cos(hypo_params.track_zenith)
+
+    #create array at vertex
+    photon_array = np.array([t, x, y, z, photons_per_segment, dir_x * 0.562, dir_y * 0.562, dir_z * 0.562])
+    photon_array = photon_array[np.newaxis]
+
+    #loop through track, appending new photon dump on axis 0
+    #loop uses rest mass of 105.658 MeV/c^2 for muon
+    rest_mass = 0.105658
+    while track_energy > rest_mass:
```
Comment (jllanfranchi) [inline retro/discrete_muon_kernels.py]: `while True:`

Comment (jllanfranchi) [inline retro/discrete_muon_kernels.py]: `while True`

File: retro/discrete_muon_kernels.py [hunk 4]
```diff
@@ -118,3 +120,109 @@ def const_energy_loss_muon(hypo_params, limits=None, dt=1.0):
...
+        z += dt * dir_z * SPEED_OF_LIGHT_M_PER_NS
+
+        #change energy of muon
+        dedx = interpolate.splev(track_energy, spline, der=0)
+        track_energy -= dedx * segment_length
+
+        #append new row if energy still above rest mass
+        if track_energy > rest_mass:
+            photon_array = np.append(photon_array, np.array([t, x, y, z, photons_per_segment, dir_x * 0.562, dir_y * 0.562, dir_z * 0.562])[np.newaxis], axis=0)
+
```
Comment (jllanfranchi) [inline retro/discrete_muon_kernels.py]:
```python
else:
    break
```

**Review by jllanfranchi — COMMENTED** [review]
What is DiscrtsHypo?

**Comment by kjc5697** [comment]
discrtsHypo is the original function I wrote before I converted to using hypo_params.  We don't need it anymore

**Comment by jllanfranchi** [comment]
Ah, one more thing I caught: put your CSV file into the `data` directory and update your reference to it accordingly.

**Comment by jllanfranchi** [comment]
Looks good!

**Pull request merged**